incubator/binding-grpc-kafka/src/main/java/io/aklivity/zilla/runtime/binding/grpc/kafka/internal/stream/GrpcKafkaProxyFactory.java

```diff
@@ -76,153 +86,971 @@ public final class GrpcKafkaProxyFactory implements GrpcKafkaStreamFactory
+            final Array32FW<KafkaOffsetFW> partitions = kafkaMergedBeginEx != null ? kafkaMergedBeginEx.partitions() : null;
+
+            if (kafkaMergedBeginEx != null)
+            {
+                String8FW encodedProgress = messageField.encodeProgressOnly(partitions);
+                Varuint32FW len = lenRW.set(encodedProgress.length()).build();
+                int lenSize = len.sizeof();
+                replyPad = result.lastMessageId().sizeof() + lenSize + encodedProgress.length();
+            }
+
+            delegate.onKafkaBegin(traceId, authorization, extension);
+        }
+
+        private void onKafkaData(
+            DataFW data)
+        {
+            final long sequence = data.sequence();
+            final long acknowledge = data.acknowledge();
+            final long traceId = data.traceId();
+            final long authorization = data.authorization();
+            final long budgetId = data.budgetId();
+            final int reserved = data.reserved();
+
+            assert acknowledge <= sequence;
+            assert sequence >= replySeq;
+
+            replySeq = sequence + reserved;
+
+            assert replyAck <= replySeq;
+
+            if (replySeq > replyAck + replyMax)
+            {
+                doKafkaReset(traceId, authorization);
+                delegate.onKafkaReset(traceId, authorization);
+            }
+            else
+            {
+                final int flags = data.flags();
+                final OctetsFW payload = data.payload();
+                final OctetsFW extension = data.extension();
+
+                final MutableDirectBuffer encodeBuffer = writeBuffer;
+                final int encodeOffset = DataFW.FIELD_OFFSET_PAYLOAD;
+                final int payloadSize = payload.sizeof();
+
+                if ((flags & DATA_FLAG_INIT) != 0x00)
+                {
+                    final ExtensionFW dataEx = extension.get(extensionRO::tryWrap);
+                    final KafkaDataExFW kafkaDataEx =
+                        dataEx != null && dataEx.typeId() == kafkaTypeId ? extension.get(kafkaDataExRO::tryWrap) : null;
+                    final KafkaMergedDataExFW kafkaMergedDataEx =
+                        kafkaDataEx != null && kafkaDataEx.kind() == KafkaDataExFW.KIND_MERGED ? kafkaDataEx.merged() : null;
+                    final Array32FW<KafkaOffsetFW> progress = kafkaMergedDataEx != null ? kafkaMergedDataEx.progress() : null;
+
+                    encodedId = messageField.encodeProgressOnly(progress);
+                }
+
+                int encodeProgress = encodeOffset;
+
+                encodeBuffer.putBytes(encodeProgress, payload.buffer(), payload.offset(), payloadSize);
+                encodeProgress += payloadSize;
+
+                if ((flags & DATA_FLAG_FIN) != 0x00) // FIN
+                {
+                    Varuint32FW key = result.lastMessageId();
```

Comment:
Suggest renaming `key` to `messageId` and `result.lastMessageId()` to `result.messageId()`.